README.md
# NAME

GD::Graph::Polar - Perl package to create polar graphs using GD package

# SYNOPSIS

    use GD::Graph::Polar;
    my $obj = GD::Graph::Polar->new(size=>480, radius=>100);
    $obj->addPoint(50=>25);                                  #radius => angle (e.g. polar form of complex number notation)
    $obj->addPoint_rad(50=>3.1415);
    $obj->addGeoPoint(75=>25);
    $obj->addGeoPoint_rad(75=>3.1415);
    $obj->addLine($radius0=>$theta0, $radius1=>$theta1);
    $obj->addLine_rad($radius0=>$theta0, $radius1=>$theta1);
    $obj->addGeoLine($radius0=>$theta0, $radius1=>$theta1);
    $obj->addGeoLine_rad($radius0=>$theta0, $radius1=>$theta1);
    $obj->addArc($radius0=>$theta0, $radius1=>$theta1);
    $obj->addArc_rad($radius0=>$theta0, $radius1=>$theta1);
    $obj->addGeoArc($radius0=>$theta0, $radius1=>$theta1);
    $obj->addGeoArc_rad($radius0=>$theta0, $radius1=>$theta1);
    $obj->addString($radius=>$theta, "Hello World!");
    $obj->addString_rad($radius=>$theta, "Hello World!");
    $obj->addGeoString($radius=>$theta, "Hello World!");
    $obj->addGeoString_rad($radius=>$theta, "Hello World!");
    $obj->font(gdSmallFont);  #sets the current font from GD exports
    $obj->color("blue");      #sets the current color from Graphics::ColorNames
    $obj->color([0,0,0]);     #sets the current color [red,green,blue]
    print $obj->draw; #PNG image

# DESCRIPTION

This package is a wrapper around GD to produce polar graphs with an easy interface.  I use this package to plot antenna patterns on a graph with data from the [RF::Antenna::Planet::MSI::Format](https://metacpan.org/pod/RF::Antenna::Planet::MSI::Format) package.

# CONSTRUCTOR

## new

The new constructor. 

    my $obj = GD::Graph::Polar->new(           #default values
                                    size          => 480,    #width and height in pixels
                                    radius        => 1,      #max value of the radius
                                    radius_origin => 0,      #value at the origin
                                    ticks         => 10,     #number of major ticks
                                    border        => 2,      #pixel border around graph
                                    rgbfile       => "/usr/X11R6/lib/X11/rgb.txt"
                                   );

# METHODS

## addPoint

Method to add a point to the graph.

    $obj->addPoint(50=>25);

## addPoint\_rad

Method to add a point to the graph.

    $obj->addPoint_rad(50=>3.1415);

## addGeoPoint

Method to add a point to the graph.

    $obj->addGeoPoint(75=>25);

## addGeoPoint\_rad

Method to add a point to the graph.

    $obj->addGeoPoint_rad(75=>3.1415);

## addLine

Method to add a line to the graph.

    $obj->addLine(50=>25, 75=>35);

## addLine\_rad

Method to add a line to the graph.

    $obj->addLine_rad(50=>3.14, 75=>3.45);

## addGeoLine

Method to add a line to the graph.

    $obj->addGeoLine(50=>25, 75=>35);

## addGeoLine\_rad

Method to add a line to the graph.

    $obj->addGeoLine_rad(50=>3.14, 75=>3.45);

## addArc

Method to add an arc to the graph.

    $obj->addArc(50=>25, 75=>35);

## addArc\_rad

Method to add an arc to the graph.

    $obj->addArc_rad(50=>3.14, 75=>3.45);

## addGeoArc

Method to add an arc to the graph.

    $obj->addGeoArc(50=>25, 75=>35);

## addGeoArc\_rad

Method to add an arc to the graph.

    $obj->addGeoArc_rad(50=>25, 75=>35);

## addString

Method to add a string to the graph.

## addString\_rad

Method to add a string to the graph.

## addGeoString

Method to add a string to the graph.

## addGeoString\_rad

Method to add a string to the graph.

# Objects

## gdimage

Returns a [GD](https://metacpan.org/pod/GD) object

## gcnames

Returns a [Graphics::ColorNames](https://metacpan.org/pod/Graphics::ColorNames) object.

# Properties

## color

Method to set or return the current drawing color

    my $colorobj = $obj->color("blue");     #if Graphics::ColorNames available
    my $colorobj = $obj->color([77,82,68]); #rgb=>[decimal,decimal,decimal]
    my $colorobj = $obj->color;

Default: \[0,0,0\] (i.e., black)

## font

Method to set or return the current drawing font (only needed by the very few)

    use GD qw(gdGiantFont gdLargeFont gdMediumBoldFont gdSmallFont gdTinyFont);
    $obj->font(gdSmallFont); #the default
    $obj->font;

Default: gdSmallFont

## size

Sets or returns the width and height of the image in pixels.

Default: 480

## radius

Sets or returns the radius of the graph which sets the scale of the maximum value of the graph.

Default: 1

## radius\_origin

Sets or returns the radius origin of the graph which sets the value scale at the origin of the graph.

Default: 0

## border

Sets and returns the number of pixels that border the graph on the image.

Default: 2

## ticks

Sets and returns the number of ticks on the graph.

Default: 10

## axes

Sets and returns the number of axes (plural of axis) on the graph.

Default: 4

## rgbfile

Sets or returns an RGB file.

Note: This method will search in a few locations for a file.

## draw

Method returns a PNG binary blob.

    my $png_binary = $obj->draw;

# LICENSE

MIT License

Copyright (c) 2022 Michael R. Davis

# SEE ALSO

[GD](https://metacpan.org/pod/GD), [Geo::Constants](https://metacpan.org/pod/Geo::Constants), [Geo::Functions](https://metacpan.org/pod/Geo::Functions), [Graphics::ColorNames](https://metacpan.org/pod/Graphics::ColorNames)


Changes
Changes for package GD-Graph-Polar

0.22 2022-10-18
  - Added axes property
  - Removed use of $_ in favor of named loop variables

0.21 2022-10-13
  - Added radius_origin method to support display of antenna paterns with dB scale (negatives)
  - Updated _scale to support radius_origin
  - Updated all methods to use $radius => $theta instead of $r=>$t
  - Added t/002_radius_origin.t as author test

0.20 2022-09-22
  - Updated rbgfile method to fixed issue calling Graphics::ColorNames which requires files to be true according to File::Spec::Functions::file_name_is_absolute - GitHub #1 - Slaven Rezić

0.19 2022-09-20
  - Fixed tests logic for CPAN testers - GitHub #1 - Slaven Rezić
  - Updated Documentation

0.18 2022-09-20
  - Changed license to MIT
  - Added GitHub repo information
  - Cleaned up code a little

0.17 2012-05-16
  - Lazy load all of the stuff in the initialize method

0.16 2011-03-27
  - Added Todo
  - Added perl-GD-Graph-Polar.spec to distribution
  - Added Tests for kwalitee, changes, pods to svn not dist
  - Moved scripts from bin to scripts folder
  - Renamed CHANGES to Changes to pass tests
  - Updated format of Changes to pass tests
  - Dropped MANIFEST.SKIP as I no longer use make manifest
  - Dropped META.yml as CPAN now auto generates it

0.15_2 2010-06-10
  - Updated tests
  - Dropped use vars

0.15 2006-03-14
  - Documentation

0.14 2006-03-13
  - Documentation

0.13 2006-02-25
  - Fixed calls using named colors to the color method failing

0.12 2006-02-25
  - Graphics::ColorNames is optional e.g. black=>[0,0,0] or black=>"black"

0.11 2006-02-17
  - Modification to account for rbg.txt file is not in a standard location

0.10 2006-02-01
  - Parameters ticks and border can now be 0

0.09 2006-01-25
  - Corrected Error in _scale method
  - Corrected Error in _imgxy_xy method
  - Corrected Error in addGeoString method

0.08 2006-01-24
  - Added Geo equivalent functions

0.07 2006-01-24
  - Added color, font, and addString methods
  - Added example-text.pl

0.06 2006-01-16
  - Documentation

0.05 2006-01-13
  - Added addGeoLine methods

0.04 2006-01-13
  - Added addArc methods

0.03 2006-01-04
  - Documentation

0.02 2006-12-22
  - Renamed image method to draw
  - Added addLine method

0.01 2006-12-20
  - Original version


Makefile.PL
use ExtUtils::MakeMaker;
WriteMakefile(
  NAME          => 'GD::Graph::Polar',
  VERSION_FROM  => 'lib/GD/Graph/Polar.pm',
  ABSTRACT_FROM => 'lib/GD/Graph/Polar.pm',
  PREREQ_PM     => {
                    'GD'                => 0,
                    'Geo::Constants'    => 0.04,
                    'Geo::Functions'    => 0.03,
                    'Test::Simple'      => 0.44, #test req
                    'List::Util'        => 0,
                    'Package::New'      => 0,
                    'Path::Class'       => 0,    #tests
                   #'Graphics::ColorNames'=>0, #no longer required but nice to have
                   },
  'META_MERGE' => {
                   'resources' => {
                                   'repository' => {
                                                    'web' => 'https://github.com/mrdvt92/perl-GD-Graph-Polar.git',
                                                    'url' => 'git@github.com:mrdvt92/perl-GD-Graph-Polar.git',
                                                    'type' => 'git'
                                                   },
                                   'homepage' => 'https://github.com/mrdvt92/perl-GD-Graph-Polar',
                                   'bugtracker' => {
                                                    'web' => 'https://github.com/mrdvt92/perl-GD-Graph-Polar/issues'
                                                   },
                                  },
                   'meta-spec' => {
                                   'version' => 2
                                  },
                  },
             );


lib/GD/Graph/Polar.pm
package GD::Graph::Polar;
use strict;
use warnings;
use base qw{Package::New};
use Cwd qw{};
use Geo::Constants qw{PI};
use Geo::Functions qw{rad_deg};
use GD qw{gdSmallFont};

our $VERSION = '0.21';

=head1 NAME

GD::Graph::Polar - Perl package to create polar graphs using GD package

=head1 SYNOPSIS

  use GD::Graph::Polar;
  my $obj = GD::Graph::Polar->new(size=>480, radius=>100);
  $obj->addPoint(50=>25);                                  #radius => angle (e.g. polar form of complex number notation)
  $obj->addPoint_rad(50=>3.1415);
  $obj->addGeoPoint(75=>25);
  $obj->addGeoPoint_rad(75=>3.1415);
  $obj->addLine($radius0=>$theta0, $radius1=>$theta1);
  $obj->addLine_rad($radius0=>$theta0, $radius1=>$theta1);
  $obj->addGeoLine($radius0=>$theta0, $radius1=>$theta1);
  $obj->addGeoLine_rad($radius0=>$theta0, $radius1=>$theta1);
  $obj->addArc($radius0=>$theta0, $radius1=>$theta1);
  $obj->addArc_rad($radius0=>$theta0, $radius1=>$theta1);
  $obj->addGeoArc($radius0=>$theta0, $radius1=>$theta1);
  $obj->addGeoArc_rad($radius0=>$theta0, $radius1=>$theta1);
  $obj->addString($radius=>$theta, "Hello World!");
  $obj->addString_rad($radius=>$theta, "Hello World!");
  $obj->addGeoString($radius=>$theta, "Hello World!");
  $obj->addGeoString_rad($radius=>$theta, "Hello World!");
  $obj->font(gdSmallFont);  #sets the current font from GD exports
  $obj->color("blue");      #sets the current color from Graphics::ColorNames
  $obj->color([0,0,0]);     #sets the current color [red,green,blue]
  print $obj->draw; #PNG image

=head1 DESCRIPTION

This package is a wrapper around GD to produce polar graphs with an easy interface.  I use this package to plot antenna patterns on a graph with data from the L<RF::Antenna::Planet::MSI::Format> package.

=head1 CONSTRUCTOR

=head2 new

The new constructor. 

  my $obj = GD::Graph::Polar->new(           #default values
                                  size          => 480,    #width and height in pixels
                                  radius        => 1,      #max value of the radius
                                  radius_origin => 0,      #value at the origin
                                  ticks         => 10,     #number of major ticks
                                  border        => 2,      #pixel border around graph
                                  rgbfile       => "/usr/X11R6/lib/X11/rgb.txt"
                                 );

=head1 METHODS

=head2 addPoint

Method to add a point to the graph.

  $obj->addPoint(50=>25);

=cut

sub addPoint {
  my $self   = shift;
  my $radius = shift;
  my $theta  = rad_deg(shift());
  return $self->addPoint_rad($radius,$theta);
}

=head2 addPoint_rad

Method to add a point to the graph.

  $obj->addPoint_rad(50=>3.1415);

=cut

sub addPoint_rad {
  my $self    = shift;
  my $radius  = shift;
  my $theta   = shift;
  my ($x, $y) = $self->_imgxy_rt_rad($radius,$theta);
  my $icon    = 7;
  return $self->gdimage->arc($x,$y,$icon,$icon,0,360,$self->color);
}

=head2 addGeoPoint

Method to add a point to the graph.

  $obj->addGeoPoint(75=>25);

=cut

sub addGeoPoint {
  my $self   = shift;
  my $radius = shift;
  my $theta  = rad_deg(shift());
  return $self->addGeoPoint_rad($radius,$theta);
}

=head2 addGeoPoint_rad

Method to add a point to the graph.

  $obj->addGeoPoint_rad(75=>3.1415);

=cut

sub addGeoPoint_rad {
  my $self   = shift;
  my $radius = shift;
  my $theta  = PI()/2-shift();
  return $self->addPoint_rad($radius,$theta);
}

=head2 addLine

Method to add a line to the graph.

  $obj->addLine(50=>25, 75=>35);

=cut

sub addLine {
  my $self    = shift;
  my $radius0 = shift;
  my $theta0  = rad_deg(shift());
  my $radius1 = shift;
  my $theta1  = rad_deg(shift());
  return $self->addLine_rad($radius0=>$theta0, $radius1=>$theta1);
}

=head2 addLine_rad

Method to add a line to the graph.

  $obj->addLine_rad(50=>3.14, 75=>3.45);

=cut

sub addLine_rad {
  my $self      = shift;
  my $radius0   = shift;
  my $theta0    = shift;
  my $radius1   = shift;
  my $theta1    = shift;
  my ($x0=>$y0) = $self->_imgxy_rt_rad($radius0=>$theta0);
  my ($x1=>$y1) = $self->_imgxy_rt_rad($radius1=>$theta1);
  return $self->gdimage->line($x0, $y0, $x1, $y1, $self->color);
}

=head2 addGeoLine

Method to add a line to the graph.

  $obj->addGeoLine(50=>25, 75=>35);

=cut

sub addGeoLine {
  my $self    = shift;
  my $radius0 = shift;
  my $theta0  = rad_deg(shift());
  my $radius1 = shift;
  my $theta1  = rad_deg(shift());
  return $self->addGeoLine_rad($radius0=>$theta0, $radius1=>$theta1);
}

=head2 addGeoLine_rad

Method to add a line to the graph.

  $obj->addGeoLine_rad(50=>3.14, 75=>3.45);

=cut

sub addGeoLine_rad {
  my $self    = shift;
  my $radius0 = shift;
  my $theta0  = PI()/2-shift();
  my $radius1 = shift;
  my $theta1  = PI()/2-shift();
  return $self->addLine_rad($radius0=>$theta0, $radius1=>$theta1);
}

=head2 addArc

Method to add an arc to the graph.

  $obj->addArc(50=>25, 75=>35);

=cut

sub addArc {
  my $self    = shift;
  my $radius0 = shift;
  my $theta0  = rad_deg(shift());
  my $radius1 = shift;
  my $theta1  = rad_deg(shift());
  return $self->addArc_rad($radius0=>$theta0, $radius1=>$theta1);
}

=head2 addArc_rad

Method to add an arc to the graph.

  $obj->addArc_rad(50=>3.14, 75=>3.45);

=cut

sub addArc_rad {
  my $self    = shift;
  my $radius0 = shift;
  my $theta0  = shift;
  my $radius1 = shift;
  my $theta1  = shift;
  my $m       = ($radius1-$radius0) / ($theta1-$theta0);
  my $inc     = 0.02; #is this good?
  my $steps   = int(($theta1-$theta0) / $inc);
  my @array   = ();
  foreach (0 .. $steps) {
    my $theta  = $_ / $steps * ($theta1-$theta0) + $theta0;
    my $radius = $radius0 + $m * ($theta-$theta0);
    push @array, [$radius=>$theta];
  } 
  my @return = ();
  foreach (1 .. $steps) {
    push @return, $self->addLine_rad(@{$array[$_-1]}, @{$array[$_]});
  }
  return \@return;
}

=head2 addGeoArc

Method to add an arc to the graph.

  $obj->addGeoArc(50=>25, 75=>35);

=cut

sub addGeoArc {
  my $self    = shift;
  my $radius0 = shift;
  my $theta0  = rad_deg(shift());
  my $radius1 = shift;
  my $theta1  = rad_deg(shift());
  return $self->addGeoArc_rad($radius0=>$theta0, $radius1=>$theta1);
}

=head2 addGeoArc_rad

Method to add an arc to the graph.

  $obj->addGeoArc_rad(50=>25, 75=>35);

=cut

sub addGeoArc_rad {
  my $self    = shift;
  my $radius0 = shift;
  my $theta0  = PI()/2-shift();
  my $radius1 = shift;
  my $theta1  = PI()/2-shift();
  return $self->addArc_rad($radius0=>$theta0, $radius1=>$theta1);
}

=head2 addString

Method to add a string to the graph.

=cut

sub addString {
  my $self   = shift;
  my $radius = shift;
  my $theta  = rad_deg(shift());
  my $string = shift;
  return $self->addString_rad($radius=>$theta, $string);
}

=head2 addString_rad

Method to add a string to the graph.

=cut

sub addString_rad {
  my $self    = shift;
  my $radius  = shift;
  my $theta   = shift;
  my $string  = shift;
  my ($x=>$y) = $self->_imgxy_rt_rad($radius=>$theta);
  return $self->gdimage->string($self->font, $x, $y, $string, $self->color);
}

=head2 addGeoString

Method to add a string to the graph.

=cut

sub addGeoString {
  my $self   = shift;
  my $radius = shift;
  my $theta  = rad_deg(shift());
  my $string = shift;
  return $self->addGeoString_rad($radius=>$theta, $string);
}

=head2 addGeoString_rad

Method to add a string to the graph.

=cut

sub addGeoString_rad {
  my $self   = shift;
  my $radius = shift;
  my $theta  = PI()/2-shift();
  my $string = shift;
  return $self->addString_rad($radius=>$theta, $string);
}

=head1 Objects

=head2 gdimage

Returns a L<GD> object

=cut

sub gdimage {
  my $self           = shift;
  $self->{'gdimage'} = shift if @_; #set a base chart or watermark
  unless ($self->{'gdimage'}) {
    $self->{'gdimage'} = GD::Image->new($self->size, $self->size);

    # make the background transparent and interlaced
    $self->gdimage->transparent($self->color([255,255,255]));
    $self->gdimage->interlaced('true');
    
    # Put a frame around the picture
    $self->gdimage->rectangle(0, 0, $self->size - 1, $self->size - 1, $self->color([0,0,0]));
  
    if ($self->ticks > 0) {
      $self->color([192,192,192]);
      foreach (0 .. $self->ticks) {
        my $c = $self->size / 2;
        my $r = $self->_width * $_ / $self->ticks;
        $self->gdimage->arc($c,$c,$r,$r,0,360,$self->color);
      }
    }
  
    $self->color([192,192,192]);
    $self->gdimage->line($self->size/2,
                            $self->border,
                            $self->size/2,
                            $self->size-$self->border,
                            $self->color);
    $self->gdimage->line($self->border,
                            $self->size/2,
                            $self->size-$self->border,
                            $self->size/2,
                            $self->color);

    #default to black pen color
    $self->color([0,0,0]);
  }
  return $self->{'gdimage'};
}

=head2 gcnames

Returns a L<Graphics::ColorNames> object.

=cut

sub gcnames {
  my $self = shift;
  unless (defined $self->{'gcnames'}) {
    eval 'use Graphics::ColorNames';
    if ($@) {
      die('Error: Cannot load Graphics::ColorNames');
    } else {
      my $file           = $self->rgbfile; #stringify for object support
      $self->{'gcnames'} = Graphics::ColorNames->new("$file"); #file path must be true per File::Spec::Unix::file_name_is_absolute
    }
  }
  return $self->{'gcnames'};
}

=head1 Properties

=head2 color

Method to set or return the current drawing color

  my $colorobj = $obj->color("blue");     #if Graphics::ColorNames available
  my $colorobj = $obj->color([77,82,68]); #rgb=>[decimal,decimal,decimal]
  my $colorobj = $obj->color;

Default: [0,0,0] (i.e., black)

=cut

sub color {
  my $self = shift;
  if (@_) {
    my $color = shift;
    if (ref($color) eq 'ARRAY') {
      my ($r, $g, $b) = @$color;
      $self->{'color'} = $self->{'colors'}->{$r}->{$g}->{$b}||=$self->gdimage->colorAllocate(@$color);
    } else {
      if ($self->gcnames) {
        my @rgb          = $self->gcnames->rgb($color);
        @rgb             = (0,0,0) unless scalar(@rgb) == 3;
        $self->{'color'} = $self->color(\@rgb);
      } else {
        $self->{'color'} = $self->color([0,0,0]);
      }
    }
  }
  return $self->{'color'};
}

=head2 font

Method to set or return the current drawing font (only needed by the very few)

  use GD qw(gdGiantFont gdLargeFont gdMediumBoldFont gdSmallFont gdTinyFont);
  $obj->font(gdSmallFont); #the default
  $obj->font;

Default: gdSmallFont

=cut

sub font {
  my $self        = shift;
  $self->{'font'} = shift if @_;
  $self->{'font'} = gdSmallFont unless $self->{'font'};
  return $self->{'font'};
}

=head2 size

Sets or returns the width and height of the image in pixels.

Default: 480

=cut

sub size {
  my $self        = shift;
  $self->{'size'} = shift if @_;
  $self->{'size'} = 480 unless $self->{'size'};
  return $self->{'size'};
}

=head2 radius

Sets or returns the radius of the graph which sets the scale of the maximum value of the graph.

Default: 1

=cut

sub radius {
  my $self          = shift;
  $self->{'radius'} = shift if @_;
  $self->{'radius'} = 1 unless defined $self->{'radius'};
  return $self->{'radius'};
}

=head2 radius_origin

Sets or returns the radius origin of the graph which sets the value scale at the origin of the graph.

Default: 0

=cut

sub radius_origin {
  my $self                 = shift;
  $self->{'radius_origin'} = shift if @_;
  $self->{'radius_origin'} = 0 unless defined $self->{'radius_origin'};
  return $self->{'radius_origin'};
}

=head2 border

Sets and returns the number of pixels that border the graph on the image.

Default: 2

=cut

sub border {
  my $self          = shift;
  $self->{'border'} = shift if @_;
  $self->{'border'} = 2 unless defined($self->{'border'});
  return $self->{'border'};
}

=head2 ticks

Sets and returns the number of ticks on the graph.

Default: 10

=cut

sub ticks {
  my $self         = shift;
  $self->{'ticks'} = shift if @_;
  $self->{'ticks'} = 10
    unless defined($self->{'ticks'});
  return $self->{'ticks'};
}

=head2 rgbfile

Sets or returns an RGB file.

Note: This method will search in a few locations for a file.

=cut

sub rgbfile {
  my $self           = shift;
  $self->{'rgbfile'} = shift if @_;
  unless (defined $self->{'rgbfile'}) {
    my $cwd = Cwd::getcwd();
    foreach ('/etc/X11/rgb.txt', '/usr/share/X11/rgb.txt', '/usr/X11R6/lib/X11/rgb.txt', "$cwd/rgb.txt", "$cwd/../rgb.txt") {
      next unless -r;
      $self->{'rgbfile'} = $_;
      last;
    }
  }
  return $self->{'rgbfile'};
}

=head2 draw

Method returns a PNG binary blob.

  my $png_binary = $obj->draw;

=cut

sub draw {
  my $self = shift;
  return $self->gdimage->png;
}

#=head2 _scale
#
#Method returns the parameter scaled to the image.
#
#=cut

sub _scale {
  my $self   = shift;
  my $radius = shift;
  $radius    = $self->radius_origin if $radius < $self->radius_origin; #polar graphs do not support negative values
  return $self->_width / 2 * ($radius - $self->radius_origin)/($self->radius - $self->radius_origin);
}

#=head2 _width
#
#Method returns the width of the graph.
#
#=cut

sub _width {
  my $self = shift;
  return $self->size - $self->border * 2;
}

#=head2 _imgxy_xy
#
#Method to convert xy to imgxy cordinates
#
#  $obj->addPoint_rad(50=>3.1415);
#
#=cut

sub _imgxy_xy {
  my $self = shift;
  my $x    = shift();
  my $y    = shift();
  my $sz   = $self->_width;
  $x       = $sz/2 + $x + $self->border;
  $y       = $sz/2 - $y + $self->border;
  return ($x, $y);
}

#=head2 _xy_rt_rad
#
#Method to convert polar cordinate to Cartesian cordinates.
#
#=cut

sub _xy_rt_rad {
  my $self   = shift;
  my $radius = shift;
  my $theta  = shift;
  my $x      = $radius*cos($theta);
  my $y      = $radius*sin($theta);
  return ($x, $y);
}

#=head2 _imgxy_rt_rad
#
#Method to convert polar cordinate to Cartesian cordinates.
#
#=cut

sub _imgxy_rt_rad {
  my $self   = shift;
  my $radius = shift;
  my $theta  = shift;
  my ($x,$y) = $self->_xy_rt_rad($self->_scale($radius), $theta);
  return $self->_imgxy_xy($x, $y);
}

=head1 LICENSE

MIT License

Copyright (c) 2022 Michael R. Davis

=head1 SEE ALSO

L<GD>, L<Geo::Constants>, L<Geo::Functions>, L<Graphics::ColorNames>

=cut

1;
